src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java

```diff
@@ -0,0 +1,373 @@
+import org.apache.qpid.proton.amqp.messaging.TerminusDurability;
+import org.apache.qpid.proton.amqp.messaging.TerminusExpiryPolicy;
+import org.apache.qpid.proton.engine.EndpointState;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class AmqpConnectionImpl implements AmqpConnection {
+
+  public static final String PRODUCT = "vertx-amqp-client";
+  public static final Symbol PRODUCT_KEY = Symbol.valueOf("product");
+
+  private final AmqpClientOptions options;
+  private final AtomicBoolean closed = new AtomicBoolean();
+  private final AtomicReference<ProtonConnection> connection = new AtomicReference<>();
+  private final Context context;
+
+  private final List<AmqpSender> senders = new CopyOnWriteArrayList<>();
+  private final List<AmqpReceiver> receivers = new CopyOnWriteArrayList<>();
+  /**
+   * The exception handler, protected by the monitor lock.
+   */
+  private Handler<Throwable> exceptionHandler;
+
+  AmqpConnectionImpl(Context context, AmqpClientImpl client, AmqpClientOptions options,
+    ProtonClient proton, Handler<AsyncResult<AmqpConnection>> connectionHandler) {
+    this.options = options;
+    this.context = context;
+
+    runOnContext(x -> connect(client,
+      Objects.requireNonNull(proton, "proton cannot be `null`"),
+      Objects.requireNonNull(connectionHandler, "connection handler cannot be `null`"))
+    );
+  }
+
+  private void connect(AmqpClientImpl client, ProtonClient proton,
+    Handler<AsyncResult<AmqpConnection>> connectionHandler) {
+    proton
+      .connect(options, options.getHost(), options.getPort(), options.getUsername(), options.getPassword(),
+        ar -> {
+          // Called on the connection context.
+
+          if (ar.succeeded()) {
+            if (!this.connection.compareAndSet(null, ar.result())) {
+              connectionHandler.handle(Future.failedFuture("Unable to connect - already holding a connection"));
+              return;
+            }
+
+            Map<Symbol, Object> map = new HashMap<>();
+            map.put(AmqpConnectionImpl.PRODUCT_KEY, AmqpConnectionImpl.PRODUCT);
+            if (options.getContainerId() != null) {
+              this.connection.get().setContainer(options.getContainerId());
+            }
+
+            if (options.getVirtualHost() != null) {
+              this.connection.get().setHostname(options.getVirtualHost());
+            }
+
+            this.connection.get()
+              .setProperties(map)
+              .disconnectHandler(ignored -> {
+                try {
+                  onDisconnect();
+                } finally {
+                  closed.set(true);
+                }
+              })
+              .closeHandler(x -> {
+                // Not expected closing, consider it failed
```

Comment:
The client should actually close the connection on its end when the remote close arrives, to send the appropriate 'response' close frame back to the server, before then disconnecting the socket. It used to do this but it was removed in the latest changes.

The current calling of the new exception handler seems like it can occur multiple times for the same overall issue, e.g. once on unexpected remote close frame, then again on tcp disconnect. The prior endHandler stuff handled that by nulling out the callback when first firing it.